File: social-flow-backend/src/notifications/email.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as nodemailer from 'nodemailer';
import { LoggerService } from '../shared/logger/logger.service';

export interface EmailTemplate {
  subject: string;
  html: string;
  text: string;
}

export interface NotificationEmailData {
  title: string;
  message: string;
  actionUrl?: string;
  imageUrl?: string;
}

@Injectable()
export class EmailService {
  private transporter: nodemailer.Transporter;

  constructor(
    private configService: ConfigService,
    private logger: LoggerService,
  ) {
    this.transporter = nodemailer.createTransporter({
      host: this.configService.get('app.emailHost'),
      port: this.configService.get('app.emailPort'),
      secure: false,
      auth: {
        user: this.configService.get('app.emailUser'),
        pass: this.configService.get('app.emailPassword'),
      },
    });
  }

  async sendEmail(to: string, subject: string, html: string, text?: string): Promise<void> {
    try {
      await this.transporter.sendMail({
        from: this.configService.get('app.emailFrom'),
        to,
        subject,
        html,
        text: text || this.htmlToText(html),
      });

      this.logger.logBusiness('email_sent', undefined, {
        to,
        subject,
      });
    } catch (error) {
      this.logger.logError(error, 'EmailService.sendEmail', {
        to,
        subject,
      });
      throw error;
    }
  }

  async sendWelcomeEmail(email: string, name: string): Promise<void> {
    const subject = 'Welcome to Social Flow!';
    const html = this.getWelcomeEmailTemplate(name);
    const text = this.getWelcomeEmailTextTemplate(name);

    await this.sendEmail(email, subject, html, text);
  }

  async sendVerificationEmail(email: string, token: string): Promise<void> {
    const subject = 'Verify your email address';
    const html = this.getVerificationEmailTemplate(token);
    const text = this.getVerificationEmailTextTemplate(token);

    await this.sendEmail(email, subject, html, text);
  }

  async sendPasswordResetEmail(email: string, token: string): Promise<void> {
    const subject = 'Reset your password';
    const html = this.getPasswordResetEmailTemplate(token);
    const text = this.getPasswordResetEmailTextTemplate(token);

    await this.sendEmail(email, subject, html, text);
  }

  async sendNotificationEmail(email: string, data: NotificationEmailData): Promise<void> {
    const subject = data.title;
    const html = this.getNotificationEmailTemplate(data);
    const text = this.getNotificationEmailTextTemplate(data);

    await this.sendEmail(email, subject, html, text);
  }

  async sendPaymentReceiptEmail(email: string, paymentData: any): Promise<void> {
    const subject = 'Payment Receipt';
    const html = this.getPaymentReceiptEmailTemplate(paymentData);
    const text = this.getPaymentReceiptEmailTextTemplate(paymentData);

    await this.sendEmail(email, subject, html, text);
  }

  async sendSubscriptionConfirmationEmail(email: string, subscriptionData: any): Promise<void> {
    const subject = 'Subscription Confirmation';
    const html = this.getSubscriptionConfirmationEmailTemplate(subscriptionData);
    const text = this.getSubscriptionConfirmationEmailTextTemplate(subscriptionData);

    await this.sendEmail(email, subject, html, text);
  }

  private getWelcomeEmailTemplate(name: string): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Welcome to Social Flow</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #007bff; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .button { display: inline-block; padding: 10px 20px; background: #007bff; color: white; text-decoration: none; border-radius: 5px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Welcome to Social Flow!</h1>
          </div>
          <div class="content">
            <h2>Hello ${name}!</h2>
            <p>Welcome to Social Flow, the ultimate social media platform for creators and viewers alike.</p>
            <p>Get started by:</p>
            <ul>
              <li>Uploading your first video</li>
              <li>Following your favorite creators</li>
              <li>Creating engaging posts</li>
              <li>Exploring trending content</li>
            </ul>
            <p style="text-align: center;">
              <a href="${process.env.FRONTEND_URL}/dashboard" class="button">Get Started</a>
            </p>
            <p>If you have any questions, feel free to reach out to our support team.</p>
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getWelcomeEmailTextTemplate(name: string): string {
    return `
      Welcome to Social Flow!
      
      Hello ${name}!
      
      Welcome to Social Flow, the ultimate social media platform for creators and viewers alike.
      
      Get started by:
      - Uploading your first video
      - Following your favorite creators
      - Creating engaging posts
      - Exploring trending content
      
      Visit: ${process.env.FRONTEND_URL}/dashboard
      
      If you have any questions, feel free to reach out to our support team.
      
      Best regards,
      The Social Flow Team
    `;
  }

  private getVerificationEmailTemplate(token: string): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Verify your email</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #007bff; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .button { display: inline-block; padding: 10px 20px; background: #007bff; color: white; text-decoration: none; border-radius: 5px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Verify your email address</h1>
          </div>
          <div class="content">
            <p>Please click the button below to verify your email address:</p>
            <p style="text-align: center;">
              <a href="${process.env.FRONTEND_URL}/verify-email?token=${token}" class="button">Verify Email</a>
            </p>
            <p>If the button doesn't work, you can also copy and paste this link into your browser:</p>
            <p>${process.env.FRONTEND_URL}/verify-email?token=${token}</p>
            <p>This link will expire in 24 hours.</p>
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getVerificationEmailTextTemplate(token: string): string {
    return `
      Verify your email address
      
      Please click the link below to verify your email address:
      
      ${process.env.FRONTEND_URL}/verify-email?token=${token}
      
      This link will expire in 24 hours.
      
      Best regards,
      The Social Flow Team
    `;
  }

  private getPasswordResetEmailTemplate(token: string): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Reset your password</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #dc3545; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .button { display: inline-block; padding: 10px 20px; background: #dc3545; color: white; text-decoration: none; border-radius: 5px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Reset your password</h1>
          </div>
          <div class="content">
            <p>You requested to reset your password. Click the button below to create a new password:</p>
            <p style="text-align: center;">
              <a href="${process.env.FRONTEND_URL}/reset-password?token=${token}" class="button">Reset Password</a>
            </p>
            <p>If the button doesn't work, you can also copy and paste this link into your browser:</p>
            <p>${process.env.FRONTEND_URL}/reset-password?token=${token}</p>
            <p>This link will expire in 24 hours.</p>
            <p>If you didn't request this password reset, please ignore this email.</p>
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getPasswordResetEmailTextTemplate(token: string): string {
    return `
      Reset your password
      
      You requested to reset your password. Click the link below to create a new password:
      
      ${process.env.FRONTEND_URL}/reset-password?token=${token}
      
      This link will expire in 24 hours.
      
      If you didn't request this password reset, please ignore this email.
      
      Best regards,
      The Social Flow Team
    `;
  }

  private getNotificationEmailTemplate(data: NotificationEmailData): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>${data.title}</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #007bff; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .button { display: inline-block; padding: 10px 20px; background: #007bff; color: white; text-decoration: none; border-radius: 5px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>${data.title}</h1>
          </div>
          <div class="content">
            <p>${data.message}</p>
            ${data.actionUrl ? `
              <p style="text-align: center;">
                <a href="${data.actionUrl}" class="button">View Details</a>
              </p>
            ` : ''}
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getNotificationEmailTextTemplate(data: NotificationEmailData): string {
    return `
      ${data.title}
      
      ${data.message}
      
      ${data.actionUrl ? `View Details: ${data.actionUrl}` : ''}
      
      Best regards,
      The Social Flow Team
    `;
  }

  private getPaymentReceiptEmailTemplate(paymentData: any): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Payment Receipt</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #28a745; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .receipt { background: white; padding: 20px; border-radius: 5px; margin: 20px 0; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Payment Receipt</h1>
          </div>
          <div class="content">
            <div class="receipt">
              <h3>Payment Details</h3>
              <p><strong>Amount:</strong> ${paymentData.currency.toUpperCase()} ${(paymentData.amount / 100).toFixed(2)}</p>
              <p><strong>Date:</strong> ${new Date(paymentData.createdAt).toLocaleDateString()}</p>
              <p><strong>Status:</strong> ${paymentData.status}</p>
              <p><strong>Description:</strong> ${paymentData.description}</p>
            </div>
            <p>Thank you for your payment!</p>
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getPaymentReceiptEmailTextTemplate(paymentData: any): string {
    return `
      Payment Receipt
      
      Payment Details:
      Amount: ${paymentData.currency.toUpperCase()} ${(paymentData.amount / 100).toFixed(2)}
      Date: ${new Date(paymentData.createdAt).toLocaleDateString()}
      Status: ${paymentData.status}
      Description: ${paymentData.description}
      
      Thank you for your payment!
      
      Best regards,
      The Social Flow Team
    `;
  }

  private getSubscriptionConfirmationEmailTemplate(subscriptionData: any): string {
    return `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Subscription Confirmation</title>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #007bff; color: white; padding: 20px; text-align: center; }
          .content { padding: 20px; background: #f9f9f9; }
          .subscription { background: white; padding: 20px; border-radius: 5px; margin: 20px 0; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Subscription Confirmation</h1>
          </div>
          <div class="content">
            <div class="subscription">
              <h3>Subscription Details</h3>
              <p><strong>Plan:</strong> ${subscriptionData.plan}</p>
              <p><strong>Amount:</strong> ${subscriptionData.currency.toUpperCase()} ${(subscriptionData.amount / 100).toFixed(2)}</p>
              <p><strong>Billing Cycle:</strong> ${subscriptionData.interval}</p>
              <p><strong>Next Billing Date:</strong> ${new Date(subscriptionData.currentPeriodEnd).toLocaleDateString()}</p>
            </div>
            <p>Your subscription has been activated successfully!</p>
            <p>Best regards,<br>The Social Flow Team</p>
          </div>
        </div>
      </body>
      </html>
    `;
  }

  private getSubscriptionConfirmationEmailTextTemplate(subscriptionData: any): string {
    return `
      Subscription Confirmation
      
      Subscription Details:
      Plan: ${subscriptionData.plan}
      Amount: ${subscriptionData.currency.toUpperCase()} ${(subscriptionData.amount / 100).toFixed(2)}
      Billing Cycle: ${subscriptionData.interval}
      Next Billing Date: ${new Date(subscriptionData.currentPeriodEnd).toLocaleDateString()}
      
      Your subscription has been activated successfully!
      
      Best regards,
      The Social Flow Team
    `;
  }

  private htmlToText(html: string): string {
    // Simple HTML to text conversion
    return html
      .replace(/<[^>]*>/g, '')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .trim();
  }
}

File: social-flow-backend/app/notifications/services/email_service.py
"""
Email Service for sending emails via SMTP.

Handles email verification, password reset, and notification emails.
"""

import logging
from typing import Optional
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import smtplib
import ssl

from app.core.config import settings

logger = logging.getLogger(__name__)


class EmailService:
    """Email service for sending emails via SMTP."""
    
    def __init__(self):
        self.smtp_host = settings.SMTP_HOST
        self.smtp_port = settings.SMTP_PORT
        self.smtp_user = settings.SMTP_USER
        self.smtp_password = settings.SMTP_PASSWORD
        self.smtp_tls = settings.SMTP_TLS
        self.smtp_ssl = settings.SMTP_SSL
        self.email_from = settings.EMAIL_FROM or "noreply@socialflow.com"
    
    def _create_smtp_connection(self):
        """Create SMTP connection."""
        if not self.smtp_host:
            raise ValueError("SMTP_HOST not configured")
        
        try:
            if self.smtp_ssl:
                context = ssl.create_default_context()
                server = smtplib.SMTP_SSL(self.smtp_host, self.smtp_port, context=context)
            else:
                server = smtplib.SMTP(self.smtp_host, self.smtp_port)
                if self.smtp_tls:
                    context = ssl.create_default_context()
                    server.starttls(context=context)
            
            if self.smtp_user and self.smtp_password:
                server.login(self.smtp_user, self.smtp_password)
            
            return server
        except Exception as e:
            logger.error(f"Failed to create SMTP connection: {e}")
            raise
    
    async def send_email(
        self,
        to_email: str,
        subject: str,
        body_html: str,
        body_text: Optional[str] = None
    ) -> bool:
        """Send email via SMTP."""
        try:
            # Create message
            message = MIMEMultipart("alternative")
            message["Subject"] = subject
            message["From"] = self.email_from
            message["To"] = to_email
            
            # Add text and HTML parts
            if body_text:
                part1 = MIMEText(body_text, "plain")
                message.attach(part1)
            
            part2 = MIMEText(body_html, "html")
            message.attach(part2)
            
            # Send email
            with self._create_smtp_connection() as server:
                server.sendmail(self.email_from, to_email, message.as_string())
            
            logger.info(f"Email sent successfully to {to_email}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to send email to {to_email}: {e}")
            return False
    
    async def send_verification_email(
        self,
        to_email: str,
        verification_token: str,
        username: str
    ) -> bool:
        """Send email verification email."""
        verification_url = f"{settings.FRONTEND_URL}/verify-email?token={verification_token}"
        
        subject = "Verify your Social Flow account"
        
        body_html = f"""
        <html>
          <body>
            <h2>Welcome to Social Flow, {username}!</h2>
            <p>Thank you for registering. Please verify your email address by clicking the link below:</p>
            <p><a href="{verification_url}">Verify Email Address</a></p>
            <p>If you didn't create this account, you can safely ignore this email.</p>
            <p>This link will expire in 24 hours.</p>
            <br>
            <p>Best regards,<br>The Social Flow Team</p>
          </body>
        </html>
        """
        
        body_text = f"""
        Welcome to Social Flow, {username}!
        
        Thank you for registering. Please verify your email address by visiting:
        {verification_url}
        
        If you didn't create this account, you can safely ignore this email.
        This link will expire in 24 hours.
        
        Best regards,
        The Social Flow Team
        """
        
        return await self.send_email(to_email, subject, body_html, body_text)
    
    async def send_password_reset_email(
        self,
        to_email: str,
        reset_token: str,
        username: str
    ) -> bool:
        """Send password reset email."""
        reset_url = f"{settings.FRONTEND_URL}/reset-password?token={reset_token}"
        
        subject = "Reset your Social Flow password"
        
        body_html = f"""
        <html>
          <body>
            <h2>Password Reset Request</h2>
            <p>Hi {username},</p>
            <p>We received a request to reset your password. Click the link below to create a new password:</p>
            <p><a href="{reset_url}">Reset Password</a></p>
            <p>If you didn't request a password reset, you can safely ignore this email.</p>
            <p>This link will expire in 1 hour.</p>
            <br>
            <p>Best regards,<br>The Social Flow Team</p>
          </body>
        </html>
        """
        
        body_text = f"""
        Password Reset Request
        
        Hi {username},
        
        We received a request to reset your password. Visit the link below to create a new password:
        {reset_url}
        
        If you didn't request a password reset, you can safely ignore this email.
        This link will expire in 1 hour.
        
        Best regards,
        The Social Flow Team
        """
        
        return await self.send_email(to_email, subject, body_html, body_text)
    
    async def send_welcome_email(
        self,
        to_email: str,
        username: str
    ) -> bool:
        """Send welcome email after verification."""
        subject = "Welcome to Social Flow!"
        
        body_html = f"""
        <html>
          <body>
            <h2>Welcome to Social Flow, {username}!</h2>
            <p>Your email has been verified and your account is now active.</p>
            <p>Start exploring:</p>
            <ul>
              <li>Create your first post</li>
              <li>Upload a video</li>
              <li>Follow interesting creators</li>
              <li>Customize your profile</li>
            </ul>
            <p><a href="{settings.FRONTEND_URL}/feed">Go to Social Flow</a></p>
            <br>
            <p>Best regards,<br>The Social Flow Team</p>
          </body>
        </html>
        """
        
        body_text = f"""
        Welcome to Social Flow, {username}!
        
        Your email has been verified and your account is now active.
        
        Start exploring:
        - Create your first post
        - Upload a video
        - Follow interesting creators
        - Customize your profile
        
        Visit: {settings.FRONTEND_URL}/feed
        
        Best regards,
        The Social Flow Team
        """
        
        return await self.send_email(to_email, subject, body_html, body_text)


# Global email service instance
email_service = EmailService()
